homeassistant/components/telegram_bot/__init__.py

```diff
@@ -485,7 +516,42 @@ async def async_send_telegram_message(service: ServiceCall) -> ServiceResponse:
     return True
 
 
-def initialize_bot(hass: HomeAssistant, p_config: dict) -> Bot:
+async def async_setup_entry(hass: HomeAssistant, entry: TelegramBotConfigEntry) -> bool:
+    """Create the Telegram bot from config entry."""
+    bot: Bot = await hass.async_add_executor_job(initialize_bot, hass, entry.data)
+    p_type: str = entry.data[CONF_PLATFORM]
+    platform = MODULES[p_type]
+
+    _LOGGER.debug("Setting up %s.%s", DOMAIN, p_type)
+    try:
+        receiver_service = await platform.async_setup_platform(hass, bot, entry)
+    except Exception:
+        _LOGGER.exception("Error setting up Telegram bot %s", p_type)
+        return False
+
+    if receiver_service is False:
+        _LOGGER.error("Failed to initialize Telegram bot %s", p_type)
+        return False
+
+    notify_service = TelegramNotificationService(
+        hass, bot, entry, entry.options[ATTR_PARSER]
+    )
+    entry.runtime_data = notify_service
+
+    entry.async_on_unload(entry.add_update_listener(update_listener))
+
+    return True
+
+
+async def update_listener(hass: HomeAssistant, entry: TelegramBotConfigEntry) -> None:
+    """Handle options update."""
+
+    # hass.config_entries.async_reload only reloads platforms not modules
```

Comment:
I don't understand this comment. What modules need it be reloaded?

If this is about updating the bot settings, I think we should teardown the existing bot of the config entry in `async_unload_entry`. Then we can freely create a new bot with the current settings in `async_setup_entry`.